firebase-auth/src/iosMain/kotlin/dev/gitlive/firebase/auth/auth.kt

```diff
@@ -106,19 +167,47 @@ private fun <T, R> T.throwError(block: T.(errorPointer: CPointer<ObjCObjectVar<N
     }
 }
 
-private suspend fun <T, R> T.awaitResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R {
+internal suspend fun <T, R> T.awaitResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R? {
+    val job = CompletableDeferred<R?>()
+    function { result, error ->
+        if(error != null) {
+            job.completeExceptionally(error.toException())
+        } else {
+            job.complete(result)
+        }
+    }
+    return job.await()
+}
+
+internal suspend fun <T, R> T.awaitResult(default: R, function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R {
+    val job = CompletableDeferred<R>()
+    function { result, error ->
+        if(result != null) {
+            job.complete(result)
+        } else if(error != null) {
+            job.completeExceptionally(error.toException())
+        } else {
+            job.complete(default)
+        }
+    }
+    return job.await()
+}
+
+internal suspend fun <T, R> T.awaitExpectedResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R {
```

Comment:
Which API call did you need to make this change for?

Reply:
Anything that deals with optional. The original code would not complete if the completion block returned an optional result. I ran into a few instances where the callback from the iOS SDK returned both value and error as null, which is by design since null is a valid response type. In that case, the original awaitResult would never complete and cause the coroutine calling the sdk to suspend indefinitely.

Now the distinction between `awaitResult(function)` `awaitResult(default, function)` and `awaitExpectedResult(function)` jut handless optionals differently. I guess `awaitResult(default, function)` could be removed in place of just doing a `?:` assignment afterwards. The awaitExpectedResult is there for API calls where the callback receives either `(value, null)` or `(null, error)`, but never `(null, null)`. Its essentially a force unwrap. If for some reason the api does return `(null, null)` it should finish with an error.

Reply:
Thanks, which iOS API call is it that can return a null result specifically?

Reply:
I dont fully recall, but looking at my original code:
`Auth.fetchSignInMethodsForEmail`
`User.getIdToken`
`User.unlink`

Essentially anywhere I called `awaitResult` the return type can be optional and any place I called `awaitExpectedResult` does not expect a null type (hence the Expected result)

Reply:
Can't see in the docs of those calls that they would never null in successful cases also if they did then you make the return type nullable and return the null to the client

Reply:
Well they are, I ran into it when testing or I would not have changed things.

And you can't do the latter because that makes infering the type impossible, since the callback type will always be of the form `(R?, NSError?)`, while the return type could be either R or R?, depending on whether you want the optional or non-optional result. Hence we have:
`internal suspend fun <T, R> T.awaitResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R?`
and
internal suspend fun <T, R> T.awaitExpectedResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R

Notice how one returns R? and the other R.

In the original method it was `internal suspend fun <T, R> T.awaitResult(function: T.(callback: (R?, NSError?) -> Unit) -> Unit): R`, returning R as a non-optional.

And then on the fork you made of this branch it changed to `
internal suspend fun <T, R> T.awaitResult(function: T.(callback: (R, NSError?) -> Unit) -> Unit): R` where R will always be optional since the callback is always optional. This actually results in the iOS code not compiling at the moment, because the actual methods expect a non-optional return type but they receive an optional one.

So I would suggest you keep this change, because I dont think force unwrapping the result is a better solution.

Reply:
As for the documentation, Android/Kotlin is a bit clearer wrt optionality. Here's the result for fetchSignInMethods:
https://firebase.google.com/docs/reference/kotlin/com/google/firebase/auth/SignInMethodQueryResult.html

`abstract MutableList<String!>? | getSignInMethods()`

Reply:
I commited to that branch to switch from windows to mac to finish this off as it was too hard guessing the correct calls 😆 

Anyway finished it now - you can still have a single awaitResult function for both R and R? by [casting the return type](https://github.com/GitLiveApp/firebase-kotlin-sdk/pull/92/commits/74c13fce3c5ec4ad53f39a3aa91156c6b74e27c1#diff-d29e748eb38c42986c332fcbf77831123ae810cf836b06e59de2ee42e2acddb0R137)